CIPs/0008.md

```diff
@@ -0,0 +1,76 @@
+# CIP 0008: Expand Metadata to general off-chain storage
+
+- Date: 2020-05-05
+- Author: @nambrot
+- Status: WIP
+
+## Overview
+
+In CIP3, the Celo protocol has started with a simple off-chain storage mechanism mainly focused on claims an account can make about their off-chain identity and communicate to other participants in the network. Over time, there have been additional use cases which have similar access patterns that prompted the creation of `Metadata` in the first place. I.e. data that does not require the expensive cost of "hard consensus", instead just needs to be authenticated by a single party.
+
+Examples include:
+- Transaction metadata such as comments or encrypted messaging in general
+- Profile information such as name or a profile picture
+- Unidirectional attestations like an audit or badges won in The Great Celo Stake Off
+- Signed statements/claims by a party for relase binaries/docker images/package hashes
+
+## Goals
+
+- Expand storage and indexing capabilities of this off-chain data
+- Encode the notion of decryption keys
+- Lightweight messaging system
+
+
+## Proposed Solution
+
+We propose to add another claim, called the `STORAGE` claim:
+```json
+{
+  "type": "STORAGE",
+  "timestamp": 1588886114907,
+  "address": "https://example.com/celo_storage"
+}
+```
+
+Alternatively, this could be set directly on the `Accounts` smart contract directly, causing off-chain storage to effectively replace the existing metadata functionality. This would have the benefit of not duplicating similar functionality.
```

Comment:
one slightly different approach: the metadata contains a separate entry for each external object eg `{ type: storage, subtype: 'badge', url: .., hash: ... }`. This doesn't scale if you wanted to use it for things like transaction metadata, but it does nicely make the objects unforgeable and also allows each to be hosted separately - e.g as an individual gist rather than a tree.

Reply:
>  it does nicely make the objects unforgeable

What do you mean by that?

> also allows each to be hosted separately

What use cases does this benefit? I expect that some of the benefit of this CIP is to tradeoff somewhat increased upfront setup with more marginal returns

Reply:
> > it does nicely make the objects unforgeable
> 
> What do you mean by that?

as it stands this scheme includes no unified way of determining if the contents have been changed.

Reply:
Yeah that is a good point, it's related to @nategraf's point that revocation is impossible in the current system, similar to detecting if contents have changed. Either of those applications have a temporal order in common, and I believe once an application requires that, it starts to become unsuitable for this approach?